keps/sig-api-machinery/NNNN-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,739 @@
+
+Client-go is also the library used by all components, any change or regression there can have a great impact on the project.
+
+The implementation of this feature should not modify any of current behaviors and completely pluggable, so users can opt-in safely and the project can be sure there will not create any regression.
+
+### Risks and Mitigations
+
+The feature will be opt-in, users can disable it and keep having the same behavior.
+
+To avoid possible security issues, this feature will require HTTP2 and HTTPS in order to work.
+
+## Design Details
+
+There are two problems we have to solve:
+
+- get a list of available control-plane nodes
+- client capability to be able to connect to multiple control-planes
+
+### List of available control-plane nodes
+
+This problems can be solved in two different ways:
+
+1. Configuring on the client the list of available control-plane nodes
+
+This can be done extending the client-go configuration with an additional field that allows users to include a list of alternative control-plane nodes that the client can use.
+
+`k8s.io/client-go/rest/config.go`
+```go
+type Config struct {
+	// Host must be a host string, a host:port pair, or a URL to the base of the apiserver.
+	// If a URL is given then the (optional) Path of that URL represents a prefix that must
+	// be appended to all request URIs used to access the apiserver. This allows a frontend
+	// proxy to easily relocate all of the apiserver endpoints.
+	Host string
+  // AlternativeHosts must be a comma separated list if hosts, host:port pairs or URLs to the base of
+  // different apiservers. The client can use any of them to access the apiserver.
+	AlternativeHosts string
+```
+
+2. RFC7838 HTTP Alternative Services
+
+This specification defines a new concept in HTTP, "Alternative Services", that allows an origin server to nominate additional means of interacting with it on the network.
+
+All conformant clusters are required to publish a list of endpoints with the apiserver addresses.In the default implementation, this list of endpoints is generated by a reconcile loop that guarantees that only ready apiserver addresses are present, see https://github.com/kubernetes/kubernetes/tree/master/pkg/controlplane/reconcilers.
+
+The proposal is for API servers to have an option to enable the generation of the Alternative Services headers based on the list of Endpoints created for the `kubernetes.default` service, per example:
+
+```sh
+1103 12:30:59.066469 1558329 round_trippers.go:454] GET https://127.0.0.1:44267/api/v1/namespaces/default/pods?limit=500 200 OK in 1 milliseconds
+I1103 12:30:59.066484 1558329 round_trippers.go:460] Response Headers:
+I1103 12:30:59.066491 1558329 round_trippers.go:463]     Cache-Control: no-cache, private
+I1103 12:30:59.066502 1558329 round_trippers.go:463]     Alt-Svc: h2="10.0.0.2:6443", h2="10.0.0.3:6443", h2="10.0.0.4:6443
```

Comment:
it's not clear how this will interact with readiness.  the LB is checking every second or so, but the external clients will not.  Because of this, it can fail over to a not-ready server (hard fail and restart perhaps), and the clients will be unaware.

Reply:
the goal is not about nines of availability, is about being able to recover on a network or connectivity problem, it only fails over when a network problem is detected, the trade off is as you say that readiness is not completely followed. On another note, only ready endpoints are published, since the apiservers remove their endpoint when they are not ready.
Per example, the most critical problem this solves is when the connection goes stale, the only way to get out of the loop is because the http2 readIdleTimeout (after 30s IIRC) will raise an error because the ping frame that will cause the client to fail over.